crates/nu-command/src/filters/default.rs

```diff
@@ -101,79 +101,219 @@ impl Command for Default {
+    if matches!(default_value.item, Value::Closure { .. }) && calculated_value.is_none() {
+        let Value::Closure { val: closure, .. } = &default_value.item else {
+            unreachable!()
+        };
+        let mut closure = ClosureEval::new(engine_state, stack, *closure.clone());
+        for col in columns {
+            if let Some(val) = record.get_mut(col) {
+                if matches!(val, Value::Nothing { .. }) || (empty && val.is_empty()) {
+                    if let Some(new_value) = calculated_value {
+                        *val = new_value.clone();
+                    } else {
+                        // Run the closure, and store the result into calculated_value for later calls
+                        let new_value = closure
+                            .run_with_input(PipelineData::Empty)?
+                            .into_value(default_value.span)?;
+                        *calculated_value = Some(new_value.clone());
+                        *val = new_value.clone();
+                    }
+                }
+            } else if let Some(new_value) = calculated_value {
+                record.push(col.clone(), new_value.clone());
+            } else {
+                // Run the closure, and store the result into calculated_value for later calls
+                let new_value = closure
+                    .run_with_input(PipelineData::Empty)?
+                    .into_value(default_value.span)?;
+                *calculated_value = Some(new_value.clone());
+                record.push(col.clone(), new_value.clone());
+            }
+        }
+    // If the default value is not a closure, or we have the cached value, just set the value directly
+    } else {
+        for col in columns {
+            if let Some(val) = record.get_mut(col) {
+                if matches!(val, Value::Nothing { .. }) || (empty && val.is_empty()) {
+                    *val = calculated_value
+                        .as_ref()
+                        .unwrap_or(&default_value.item)
+                        .clone();
+                }
+            } else {
+                record.push(
+                    col.clone(),
+                    calculated_value
+                        .as_ref()
+                        .unwrap_or(&default_value.item)
+                        .clone(),
+                );
+            }
+        }
+    }
+    Ok(Value::record(record.clone(), Span::unknown()).into_pipeline_data())
+}
+
 fn default(
     engine_state: &EngineState,
     stack: &mut Stack,
     call: &Call,
     input: PipelineData,
     default_when_empty: bool,
 ) -> Result<PipelineData, ShellError> {
+    let input_span = input.span().unwrap_or_else(Span::unknown);
     let metadata = input.metadata();
-    let value: Value = call.req(engine_state, stack, 0)?;
-    let column: Option<Spanned<String>> = call.opt(engine_state, stack, 1)?;
+    let default_value: Spanned<Value> = call.req(engine_state, stack, 0)?;
```

Comment:
Values already have an internal span, there's no need to use `Spanned` with `Value`